File: datasets/dataset/lib/lua/magic-core/livescripts/teleport.lua
local ____lualib = require("lualib_bundle")
local __TS__New = ____lualib.__TS__New
local ____exports = {}
local ____definitions = require("shared.packets.definitions")
local TELEPORT = ____definitions.TELEPORT
local ____teleport = require("shared.packets.implementations.teleport")
local TELEPORT_PACKET = ____teleport.TELEPORT_PACKET
local TELEPORT_SPELL_PACKET = ____teleport.TELEPORT_SPELL_PACKET
function ____exports.teleport(events)
    events.CustomPacket:OnReceive(
        TELEPORT,
        function(opcode, packet, player)
            if not player then
                return
            end
            if not player:HasAura(80908) then
                return
            end
            local parsed = __TS__New(TELEPORT_PACKET)
            parsed:Read(packet)
            repeat
                local ____switch6 = parsed:getLocation()
                local ____cond6 = ____switch6 == 1
                if ____cond6 then
                    player:Teleport(
                        0,
                        -8833.379883,
                        628.627991,
                        94.006599,
                        1.06535
                    )
                    break
                end
                ____cond6 = ____cond6 or ____switch6 == 2
                if ____cond6 then
                    player:Teleport(
                        0,
                        -4918.879883,
                        -940.406006,
                        501.563995,
                        5.42347
                    )
                    break
                end
                ____cond6 = ____cond6 or ____switch6 == 3
                if ____cond6 then
                    player:Teleport(
                        1,
                        9949.55957,
                        2284.209961,
                        1341.394165,
                        1.59587
                    )
                    break
                end
                ____cond6 = ____cond6 or ____switch6 == 4
                if ____cond6 then
                    player:Teleport(
                        1,
                        1629.359985,
                        -4373.390137,
                        31.256399,
                        3.54839
                    )
                    break
                end
                ____cond6 = ____cond6 or ____switch6 == 5
                if ____cond6 then
                    player:Teleport(
                        0,
                        1584.069946,
                        241.987,
                        -52.1534,
                        0.049647
                    )
                    break
                end
                ____cond6 = ____cond6 or ____switch6 == 6
                if ____cond6 then
                    player:Teleport(
                        1,
                        -1277.369995,
                        124.804001,
                        131.287003,
                        5.22274059587
                    )
                    break
                end
            until true
        end
    )
    events.Spell:OnApply(
        80908,
        function(spell)
            local player = ToPlayer(spell:GetCaster())
            if not player then
                return
            end
            local packet = __TS__New(TELEPORT_SPELL_PACKET)
            packet:setState(1)
            packet:Write():SendToPlayer(player)
        end
    )
    events.Spell:OnRemove(
        80908,
        function(spell)
            local player = ToPlayer(spell:GetCaster())
            if not player then
                return
            end
            local packet = __TS__New(TELEPORT_SPELL_PACKET)
            packet:setState(0)
            packet:Write():SendToPlayer(player)
        end
    )
end
return ____exports

File: livescripts/teleport.ts
import { TELEPORT } from "../shared/packets/definitions";
import { TELEPORT_PACKET, TELEPORT_SPELL_PACKET } from "../shared/packets/implementations/teleport";

export function teleport(events: TSEvents) {
    events.CustomPacket.OnReceive(TELEPORT, (opcode, packet, player) => {
        if (!player) {
            return
        }

        if(!player.HasAura(UTAG('magic-core', 'SCROLL_OF_KNOWLEDGE_AURA'))){
            return
        }

        let parsed = new TELEPORT_PACKET();
        parsed.Read(packet)

        switch (parsed.getLocation()) {
            case 1.0: // Stormwind
                player.Teleport(0, -8833.379883, 628.627991, 94.006599, 1.065350)
                break;
            case 2.0: // Ironforge
                player.Teleport(0, -4918.879883, -940.406006, 501.563995, 5.423470)
                break;
            case 3.0: // Darnassus
                player.Teleport(1, 9949.559570, 2284.209961, 1341.394165, 1.595870)
                break;
            case 4.0: // Orgrimmar
                player.Teleport(1, 1629.359985, -4373.390137, 31.256399, 3.548390)
                break;
            case 5.0: // Undercity
                player.Teleport(0, 1584.069946, 241.987000, -52.153400, 0.049647)
                break;
            case 6.0: // Thunder Bluff
                player.Teleport(1, -1277.369995, 124.804001, 131.287003, 5.222740595870)
                break;
        }
    })

    events.Spell.OnApply(UTAG('magic-core', 'SCROLL_OF_KNOWLEDGE_AURA'), (spell) => {
        const player = ToPlayer(spell.GetCaster())

        if (!player) {
            return
        }

        let packet = new TELEPORT_SPELL_PACKET()
        packet.setState(1)
        packet.Write().SendToPlayer(player)
    })

    events.Spell.OnRemove(UTAG('magic-core', 'SCROLL_OF_KNOWLEDGE_AURA'), (spell) => {
        const player = ToPlayer(spell.GetCaster())

        if (!player) {
            return
        }

        let packet = new TELEPORT_SPELL_PACKET()
        packet.setState(0)
        packet.Write().SendToPlayer(player)
    })
}
